optuna/samplers/_base.py

```diff
@@ -137,6 +140,37 @@ def sample_independent(
 
         raise NotImplementedError
 
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
```

Comment:
Just accepting a `FrozenTrial` with its state and values set would speak for a cleaner API. This is however not possible since to the storage cache mechanism assumes that a finished trial is never modified.